### PR TITLE
Added L2 Normalisation operation

### PR DESCRIPTION
**Pull request by shubhamugare**

### CODE REVIEW

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp [hunk 1]
```diff
@@ -954,6 +954,64 @@ void Maxpool(MYINT *A, MYINT *B, MYINT N, MYINT H, MYINT W, MYINT C, MYINT FH, M
 	return;
 }
 
+// A = Normalise(A)
+void NormaliseL2(MYINT* A, MYINT N, MYINT H, MYINT W, MYINT C, MYINT scaleA, MYINT shrA) {
```
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.cpp:958]: name this L2Norm4 (add number of dimensions?)

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp [hunk 2]
```diff
@@ -954,6 +954,64 @@ void Maxpool(MYINT *A, MYINT *B, MYINT N, MYINT H, MYINT W, MYINT C, MYINT FH, M
...
+			for (MYITE w = 0; w < W; w++) {
+		 
+
+				// calculate the sum square
+				int32_t sumSquare = 0;
+				MYINT shrAdiv = (1<<shrA);
+
+				for (MYITE c = 0; c < C; c++) {
+						MYINT tmp = (A[n * H * W * C + h * W * C + w * C + c] / shrAdiv);
+						sumSquare += tmp*tmp;
```
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.cpp]: this is a much less accurate way for integer division. will give problems in 8 bit code. 
can you try this: 

`int64_t tmp = (A[n * H * W * C + h * W * C + w * C + c]);
sumSquare += (MYINT)((tmp*tmp) / shrAdiv)`

or you can use something like matrix multiplication in this file (which uses the FASTAPPROX flag)

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp [hunk 3]
```diff
@@ -954,6 +954,64 @@ void Maxpool(MYINT *A, MYINT *B, MYINT N, MYINT H, MYINT W, MYINT C, MYINT FH, M
...
+				}
+				
+
+				// calculate the inverse square roor of sumSquare
```
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.cpp]: root

File: Tools/SeeDot/seedot/Predictor/library_fixed.h [hunk 4]
```diff
@@ -243,6 +243,18 @@ void Relu2D(MYINT *A, MYINT H, MYINT W);
  */
 void Maxpool(MYINT *A, MYINT *B, MYINT N, MYINT H, MYINT W, MYINT C, MYINT FH, MYINT FW, MYINT strideH, MYINT strideW, MYINT HPADL, MYINT HPADR, MYINT WPADL, MYINT WPADR);
 
+
+/**
+ * Dimensions:	A is a tensor. dim(A) = [N][H][W][C]
+ * 				scaleA, shrA are integers
+ * 
+ * Exponentiation
+ * For rach channel computes the L2 norm of all its elements. And divides each number in that channel by the norm.
```
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.h]: each

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp (hunk 1 above)
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.cpp:958]: wasnt there an epsilon as well for division stability? could help if the vector is very close to zero

File: Tools/SeeDot/seedot/Predictor/library_fixed.h [hunk 5]
```diff
@@ -1042,6 +1054,62 @@ void Maxpool(TypeA* A, TypeB* B, MYINT N, MYINT H, MYINT W, MYINT C, MYINT FH, M
...
+	for (MYITE n = 0; n < N; n++) {
+		for (MYITE h = 0; h < H; h++) {
+			for (MYITE w = 0; w < W; w++) {
+
+				// calculate the sum square
+				int32_t sumSquare = 0;
+				MYINT shrAdiv = (1<<shrA);
+
+				for (MYITE c = 0; c < C; c++) {
+						MYINT tmp = (A[n * H * W * C + h * W * C + w * C + c] / shrAdiv);
```
Comment (aayan636) [inline Tools/SeeDot/seedot/Predictor/library_fixed.h]: same comment as in library_fixed.cpp
use the example of matmul in this file for multiplication

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp (hunk 1 above)
Comment (shubhamugare) [reply Tools/SeeDot/seedot/Predictor/library_fixed.cpp:958]: That would mean calculating L2 Norm. But we calculated the norm and then divide the numbers by norm. We can just add the dimension name if in future we add some other dimension.

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp (hunk 2 above)
Comment (shubhamugare) [reply Tools/SeeDot/seedot/Predictor/library_fixed.cpp]: done

File: Tools/SeeDot/seedot/Predictor/library_fixed.cpp (hunk 3 above)
Comment (shubhamugare) [reply Tools/SeeDot/seedot/Predictor/library_fixed.cpp]: done

File: Tools/SeeDot/seedot/Predictor/library_fixed.h (hunk 4 above)
Comment (shubhamugare) [reply Tools/SeeDot/seedot/Predictor/library_fixed.h]: done

**Pull request merged**